pymatgen/analysis/diffraction/tem.py

```diff
@@ -0,0 +1,523 @@
+            electron_scattering_factors[atom.symbol] = scattering_factors_for_atom
+            scattering_factors_for_atom = {}
+        return electron_scattering_factors
+
+    def cell_scattering_factors(self, structure: Structure, bragg_angles: Dict[Tuple[int, int, int], float]) \
+            -> Dict[Tuple[int, int, int], int]:
+        """
+        Calculates the scattering factor for the whole cell.
+        Args:
+            structure (Structure): The input structure.
+            bragg_angles (dict of 3-tuple to float): The Bragg angles for each hkl plane.
+        Returns:
+            dict of hkl plane (3-tuple) to scattering factor (in angstroms).
+        """
+        cell_scattering_factors = {}
+        electron_scattering_factors = self.electron_scattering_factors(structure, bragg_angles)
+        scattering_factor_curr = 0
+        for plane in bragg_angles:
+            for site in structure:
+                for sp, occu in site.species.items():
+                    g_dot_r = np.dot(np.array(plane), np.transpose(site.frac_coords))
+                    scattering_factor_curr += electron_scattering_factors[sp.symbol][plane] * np.exp(
+                        2j * np.pi * g_dot_r)
+            cell_scattering_factors[plane] = scattering_factor_curr
+            scattering_factor_curr = 0
+        return cell_scattering_factors
+
+    def cell_intensity(self, structure: Structure, bragg_angles: Dict[Tuple[int, int, int], float]) \
+            -> Dict[Tuple[int, int, int], float]:
+        """
+        Calculates cell intensity for each hkl plane. For simplicity's sake, take I = |F|**2.
+        Args:
+            structure (Structure): The input structure.
+            bragg_angles (dict of 3-tuple to float): The Bragg angles for each hkl plane.
+        Returns:
+            dict of hkl plane to cell intensity
+        """
+        csf = self.cell_scattering_factors(structure, bragg_angles)
+        plane = bragg_angles.keys()
+        csf_val = np.array(list(csf.values()))
+        cell_intensity_val = (csf_val * csf_val.conjugate()).real
+        cell_intensity = dict(zip(plane, cell_intensity_val))
+        return cell_intensity
+
+    def get_pattern(self, structure: Structure, scaled: bool = None, two_theta_range: Tuple[float, float] = None) \
+            -> pd.DataFrame:
+        """
+            Returns all relevant TEM DP info in a pandas dataframe.
+            Args:
+                structure (Structure): The input structure.
+                scaled (boolean): Required value for inheritance, does nothing in TEM pattern
+                two_theta_range (Tuple): Required value for inheritance, does nothing in TEM pattern
+            Returns:
+                PandasDataFrame
+        """
+        points = self.generate_points(-10, 11)
+        TEM_dots = self.tem_dots(structure, points)
+        field_names = ["Pos", "(hkl)", "Intnsty (norm)", "Film rad", "Interplanar Spacing"]
```

Comment:
I'd write out these in full, e.g. "Intensity", "Position", "Film radius" ... abbreviating doesn't help much.

Also `TEM_dots` (capitalized) is a PEP8 violation.

Below, why the line `dict1 = {}`?